File: prep/implementation/practice_iterations/two_pointers_3/squaring_a_sorted_array.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

// O(n) / O(1)
int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n;
  cin >> tt;

  while (tt--) {
    cin >> n;

    vector<int> a(n);
    for (int i = 0; i < n; i++) {
      cin >> a[i];
    }

    int left = 0, right = n - 1;
    int next_pos = n - 1;
    vector<int> res(n);

    while (left < right) {
      if (abs(a[left]) >= abs(a[right])) {
        res[next_pos--] = a[left] * a[left];
        left++;
      } else {
        res[next_pos--] = a[right] * a[right];
        right--;
      }
    }

    for (int e : res) {
      cout << e << " ";
    }
    cout << "\n";
  }

  return 0;
}

File: lc/coreprep/implementation/practice_iterations/iteration_archives/two_pointers_5/SquaringASortedArray.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

// O(n) / O(n)
public class SquaringASortedArray {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();

        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
          a[i] = in.nextInt();
        }

        int[] res = new int[n];
        int left = 0, right = n - 1;
        int nextPos = n - 1;

        while (left < right) {
          if (Math.abs(a[left]) >= Math.abs(a[right])) {
            res[nextPos--] = a[left] * a[left];
            left++;
          } else {
            res[nextPos--] = a[right] * a[right];
            right--;
          }
        }

        for (int e : res) {
          System.out.printf("%d ", e);
        }
        System.out.println();
      }
    }
  }
}